File: app/src/main/java/com/petnbu/petnbu/feed/comment/RepliesRecyclerViewAdapter.java
package com.petnbu.petnbu.feed.comment;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.support.v4.graphics.drawable.RoundedBitmapDrawable;
import android.support.v4.graphics.drawable.RoundedBitmapDrawableFactory;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.TextUtils;
import android.text.format.DateUtils;
import android.text.style.ForegroundColorSpan;
import android.text.style.StyleSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.bumptech.glide.RequestManager;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.BitmapImageViewTarget;
import com.bumptech.glide.request.target.Target;
import com.bumptech.glide.request.transition.Transition;
import com.google.android.gms.common.internal.Preconditions;
import com.petnbu.petnbu.BaseBindingViewHolder;
import com.petnbu.petnbu.GlideApp;
import com.petnbu.petnbu.GlideRequests;
import com.petnbu.petnbu.R;
import com.petnbu.petnbu.databinding.ViewLoadingBinding;
import com.petnbu.petnbu.model.LocalStatus;
import com.petnbu.petnbu.util.Utils;
import com.petnbu.petnbu.databinding.ViewCommentBinding;
import com.petnbu.petnbu.model.CommentUI;
import com.petnbu.petnbu.model.Photo;
import com.petnbu.petnbu.util.ColorUtils;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.Locale;

public class RepliesRecyclerViewAdapter extends RecyclerView.Adapter<BaseBindingViewHolder> {

    private static final int VIEW_TYPE_COMMENT = 1;
    private static final int VIEW_TYPE_LOADING = 2;

    private List<CommentUI> mComments;
    private RepliesRecyclerViewAdapter.OnItemClickListener mOnItemClickListener;
    private CommentsViewModel mCommentsViewModel;
    private GlideRequests mGlideRequests;

    private String mCommentId;
    private int mDataVersion;
    private boolean mAddLoadMore;

    public RepliesRecyclerViewAdapter(Context context, List<CommentUI> comments, String commentId,
                                      OnItemClickListener onItemClickListener, CommentsViewModel commentsViewModel) {
        Preconditions.checkNotNull(context);
        Preconditions.checkNotNull(commentsViewModel);

        mComments = comments != null ? comments : new ArrayList<>();
        mGlideRequests = GlideApp.with(context);
        mCommentsViewModel = commentsViewModel;
        mOnItemClickListener = onItemClickListener;
        mCommentId = commentId;
    }

    @NonNull
    @Override
    public BaseBindingViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == VIEW_TYPE_LOADING) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_loading, parent, false);
            return new ViewLoadingHolder(view);
        } else {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_comment, parent, false);
            return new ViewHolder(view);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull BaseBindingViewHolder holder, int position) {
        if (getItemViewType(position) != VIEW_TYPE_LOADING) {
            holder.bindData(mComments.get(position));
        }
    }

    @Override
    public void onBindViewHolder(@NonNull BaseBindingViewHolder holder, int position, @NonNull List<Object> payloads) {
        if(!payloads.isEmpty()) {
            if(getItemViewType(position) != VIEW_TYPE_LOADING)
                holder.bindData(mComments.get(position), payloads);
        } else {
            super.onBindViewHolder(holder, position, payloads);
        }
    }

    @Override
    public int getItemCount() {
        return getDataItemCount() + (mAddLoadMore ? 1 : 0);
    }

    private int getDataItemCount() {
        return mComments.size();
    }

    @Override
    public int getItemViewType(int position) {
        if (position < getDataItemCount() && getDataItemCount() > 0) {
            return VIEW_TYPE_COMMENT;
        }
        return VIEW_TYPE_LOADING;
    }

    public void setAddLoadMore(boolean addLoadMore) {
        if (mAddLoadMore != addLoadMore) {
            if (addLoadMore) {
                mAddLoadMore = addLoadMore;
                notifyItemInserted(getLoadingMoreItemPosition());
            } else {
                notifyItemRemoved(getLoadingMoreItemPosition());
                mAddLoadMore = addLoadMore;
            }
        }
    }

    private int getLoadingMoreItemPosition() {
        return mAddLoadMore ? getItemCount() - 1 : RecyclerView.NO_POSITION;
    }

    @SuppressLint("StaticFieldLeak")
    public void setComments(List<CommentUI> comments) {
        mDataVersion++;
        final int startVersion = mDataVersion;
        new AsyncTask<Void, Void, DiffUtil.DiffResult>() {
            @Override
            protected DiffUtil.DiffResult  doInBackground(Void... voids) {
                return DiffUtil.calculateDiff(new RepliesRecyclerViewAdapter.CommentsDiffCallback(mComments, comments));
            }
            @Override
            protected void onPostExecute(DiffUtil.DiffResult diffResult) {
                if (startVersion != mDataVersion) {
                    // ignore update
                    return;
                }
                mComments = comments;
                diffResult.dispatchUpdatesTo(RepliesRecyclerViewAdapter.this);
            }
        }.execute();
    }

    protected class ViewHolder extends BaseBindingViewHolder<ViewCommentBinding, CommentUI> {

        private CommentUI mComment;

        public ViewHolder(View itemView) {
            super(itemView);

            mBinding.imgProfile.setOnClickListener(v -> {
                if (getAdapterPosition() != RecyclerView.NO_POSITION) {
                    CommentUI commentUI = mComments.get(getAdapterPosition());
                    if (commentUI.getLocalStatus() != LocalStatus.STATUS_UPLOADING)
                        mCommentsViewModel.openUserProfile(commentUI.getOwner().getUserId());
                }
            });

            mBinding.imgLike.setOnClickListener(v -> {
                if (getAdapterPosition() != RecyclerView.NO_POSITION) {
                    CommentUI commentUI = mComments.get(getAdapterPosition());
                    if (commentUI.getLocalStatus() != LocalStatus.STATUS_UPLOADING)
                        mCommentsViewModel.likeSubCommentClicked(commentUI.getId());
                }
            });
        }

        @Override
        public void bindData(CommentUI item) {
            mComment = item;
            displayUserInfo();
            displayContent();
            displayInfo();
        }

        @Override
        public void bindData(CommentUI item, List<Object> payloads) {
            mComment = item;

            Bundle bundle = (Bundle) payloads.get(0);
            if(bundle.getBoolean("like_status")) {
                displayLikeInfo();
            }
        }

        private void displayUserInfo() {
            String avatarUrl = TextUtils.isEmpty(mComment.getOwner().getAvatar().getThumbnailUrl()) ?
                    mComment.getOwner().getAvatar().getOriginUrl() : mComment.getOwner().getAvatar().getThumbnailUrl();
            mGlideRequests.load(avatarUrl)
                    .centerInside()
                    .into(mBinding.imgProfile);
        }

        private void displayContent() {
            SpannableStringBuilder builder = new SpannableStringBuilder(mComment.getOwner().getName() + "");
            int start = 0;
            builder.setSpan(new StyleSpan(Typeface.BOLD), start, builder.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            builder.setSpan(new ForegroundColorSpan(Color.BLACK), start, builder.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            builder.append("  ");

            start = builder.length();
            builder.append(mComment.getContent());
            builder.setSpan(new ForegroundColorSpan(ColorUtils.modifyAlpha(Color.BLACK, 0.8f)), start, builder.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            mBinding.tvContent.setText(builder);

            if (mComment.getPhoto() != null) {
                mBinding.imgPhoto.setVisibility(View.VISIBLE);
                Photo photo = mComment.getPhoto();

                float ratio = photo.getWidth() / (float) photo.getHeight();
                if (photo.getWidth() > photo.getHeight()) {
                    mBinding.imgPhoto.set(1.0f, 1/ratio);
                } else {
                    mBinding.imgPhoto.set(1.0f/2, 1/(2*ratio));
                }
                String photoUrl = TextUtils.isEmpty(mComment.getPhoto().getSmallUrl()) ?
                        mComment.getPhoto().getOriginUrl() : mComment.getPhoto().getSmallUrl();

                mGlideRequests.load(photoUrl)
                        .centerInside()
                        .into(mBinding.imgPhoto);
            } else {
                mBinding.imgPhoto.setVisibility(View.GONE);
            }
        }

        private void displayInfo() {
            mBinding.tvDate.setText(DateUtils.getRelativeTimeSpanString(mComment.getTimeCreated().getTime(),
                    Calendar.getInstance().getTimeInMillis(), 0L, DateUtils.FORMAT_ABBREV_RELATIVE));
            if (!mComment.getId().equals(mCommentId)) {
                mBinding.divider.setVisibility(View.INVISIBLE);
                displayLikeInfo();
            } else {
                mBinding.divider.setVisibility(View.VISIBLE);
                mBinding.progressBar.setVisibility(View.GONE);
            }
        }

        private void displayLikeInfo() {
            if (mComment.getLocalStatus() == LocalStatus.STATUS_UPLOADING || mComment.isLikeInProgress()) {
                mBinding.imgLike.setVisibility(View.INVISIBLE);
                mBinding.progressBar.setVisibility(View.VISIBLE);
            } else {
                mBinding.imgLike.setVisibility(View.VISIBLE);
                mBinding.progressBar.setVisibility(View.GONE);
                if(mComment.isLiked()){
                    mBinding.imgLike.setImageResource(R.drawable.ic_favorite_red_24dp);
                }else{
                    mBinding.imgLike.setImageResource(R.drawable.ic_favorite_border_black_24dp);
                }
            }
            if (mComment.getLikeCount() > 0) {
                mBinding.tvLikesCount.setVisibility(View.VISIBLE);
                mBinding.tvLikesCount.setText(String.format("%d %s", mComment.getLikeCount(),
                        mComment.getLikeCount() > 1 ? "likes" : "like"));
            } else {
                mBinding.tvLikesCount.setVisibility(View.GONE);
            }
        }
    }

    protected class ViewLoadingHolder extends BaseBindingViewHolder<ViewLoadingBinding, Void> {

        public ViewLoadingHolder(View itemView) {
            super(itemView);
        }

        @Override
        public void bindData(Void item) {
        }

        @Override
        public void bindData(Void item, List<Object> payloads) {

        }
    }

    private static class CommentsDiffCallback extends DiffUtil.Callback {

        private final List<CommentUI> oldData;
        private final List<CommentUI> newData;

        private CommentsDiffCallback(List<CommentUI> oldData, List<CommentUI> newData) {
            this.oldData = oldData;
            this.newData = newData;
        }

        @Override
        public int getOldListSize() {
            return oldData.size();
        }

        @Override
        public int getNewListSize() {
            return newData.size();
        }

        @Override
        public boolean areItemsTheSame(int oldPos, int newPos) {
            return oldData.get(oldPos).getId().equals(newData.get(newPos).getId());
        }

        @Override
        public boolean areContentsTheSame(int oldPos, int newPos) {
            return oldData.get(oldPos).equals(newData.get(newPos));
        }

        @Nullable
        @Override
        public Object getChangePayload(int oldItemPosition, int newItemPosition) {
            Bundle bundle = new Bundle();
            CommentUI oldComment = oldData.get(oldItemPosition);
            CommentUI newComment = newData.get(newItemPosition);
            if(oldComment.isLikeInProgress() != newComment.isLikeInProgress() || oldComment.isLiked() != newComment.isLiked()) {
                bundle.putBoolean("like_status", true);
                return bundle;
            }
            return super.getChangePayload(oldItemPosition, newItemPosition);
        }
    }

    public interface OnItemClickListener {

        void onPhotoClicked(Photo photo);

        void onLikeClicked(String commentId);
    }
}

File: app/src/main/java/com/petnbu/petnbu/ui/comment/RepliesRecyclerViewAdapter.kt
package com.petnbu.petnbu.ui.comment

import android.content.Context
import android.graphics.Color
import android.os.Bundle
import android.support.v7.recyclerview.extensions.ListAdapter
import android.support.v7.util.DiffUtil
import android.support.v7.widget.RecyclerView
import android.text.SpannableStringBuilder
import android.text.format.DateUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.text.bold
import androidx.core.text.color
import androidx.core.view.isInvisible
import androidx.core.view.isVisible
import com.petnbu.petnbu.GlideApp
import com.petnbu.petnbu.R
import com.petnbu.petnbu.databinding.ViewCommentBinding
import com.petnbu.petnbu.databinding.ViewLoadingBinding
import com.petnbu.petnbu.model.CommentUI
import com.petnbu.petnbu.model.Photo
import com.petnbu.petnbu.model.isUploading
import com.petnbu.petnbu.ui.BaseBindingViewHolder
import com.petnbu.petnbu.util.ColorUtils
import java.util.*

class RepliesRecyclerViewAdapter(context: Context,
                                 private val commentId: String,
                                 private val onItemClickListener: OnItemClickListener,
                                 private val commentsViewModel: CommentsViewModel)
    : ListAdapter<CommentUI, BaseBindingViewHolder<*, *>>(CommentDiffCallback()) {

    private val glideRequests = GlideApp.with(context)

    var addLoadMore: Boolean = false
        set(value) {
            if (field != value) {
                if (value) {
                    field = value
                    notifyItemInserted(loadingMoreItemPosition)
                } else {
                    notifyItemRemoved(loadingMoreItemPosition)
                    field = value
                }
            }
        }

    private val dataItemCount: Int
        get() = super.getItemCount()

    private val loadingMoreItemPosition: Int
        get() = if (addLoadMore) itemCount - 1 else RecyclerView.NO_POSITION

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseBindingViewHolder<*, *> {
        return if (viewType == VIEW_TYPE_LOADING) {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.view_loading, parent, false)
            ViewLoadingHolder(view)
        } else {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.view_comment, parent, false)
            ViewHolder(view)
        }
    }

    override fun onBindViewHolder(holder: BaseBindingViewHolder<*, *>, position: Int) {
        (holder as? ViewHolder)?.bindData(getItem(position))
    }

    override fun onBindViewHolder(holder: BaseBindingViewHolder<*, *>, position: Int, payloads: List<Any>) {
        if (!payloads.isEmpty())
            (holder as? ViewHolder)?.bindData(getItem(position), payloads)
        else
            super.onBindViewHolder(holder, position, payloads)
    }

    override fun getItemCount() = dataItemCount + if (addLoadMore) 1 else 0

    override fun getItemViewType(position: Int) = if (position < dataItemCount && dataItemCount > 0) VIEW_TYPE_COMMENT else VIEW_TYPE_LOADING

    private inner class ViewHolder(itemView: View) : BaseBindingViewHolder<ViewCommentBinding, CommentUI>(itemView) {

        private lateinit var comment: CommentUI

        init {
            mBinding.imgProfile.setOnClickListener {
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    getItem(adapterPosition).run {
                        if (!isUploading())
                            commentsViewModel.openUserProfile(owner.userId)
                    }
                }
            }

            mBinding.imgLike.setOnClickListener {
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    getItem(adapterPosition).run {
                        if (!isUploading())
                            commentsViewModel.likeSubCommentClicked(id)
                    }
                }
            }
        }

        override fun bindData(item: CommentUI) {
            comment = item
            displayUserInfo()
            displayContent()
            displayInfo()
        }

        override fun bindData(item: CommentUI, payloads: List<Any>) {
            comment = item

            (payloads[0] as? Bundle)?.run {
                if (getBoolean("like_status"))
                    displayLikeInfo()
            }
        }

        private fun displayUserInfo() {
            comment.owner.run {
                val avatarUrl = if (!avatar.thumbnailUrl.isNullOrEmpty()) avatar.thumbnailUrl else avatar.originUrl
                glideRequests.load(avatarUrl)
                        .centerInside()
                        .into(mBinding.imgProfile)
            }
        }

        private fun displayContent() {
            mBinding.tvContent.text = SpannableStringBuilder()
                    .bold {
                        color(color = Color.BLACK) {
                            append(comment.owner.name)
                        }
                    }
                    .append("  ")
                    .color(color = ColorUtils.modifyAlpha(Color.BLACK, 0.8f)) {
                        append(comment.content)
                    }

            comment.photo?.run {
                mBinding.imgPhoto.isVisible = true

                val ratio = width / height.toFloat()
                if (width > height)
                    mBinding.imgPhoto.set(1.0f, 1 / ratio)
                else
                    mBinding.imgPhoto.set(1.0f / 2, 1 / (2 * ratio))

                val photoUrl = if (!smallUrl.isNullOrEmpty()) smallUrl else originUrl
                glideRequests.load(photoUrl)
                        .centerInside()
                        .into(mBinding.imgPhoto)
            } ?: kotlin.run {
                mBinding.imgPhoto.isVisible = false
            }
        }

        private fun displayInfo() {
            mBinding.tvDate.text = DateUtils.getRelativeTimeSpanString(comment.timeCreated.time,
                    Calendar.getInstance().timeInMillis, 0L, DateUtils.FORMAT_ABBREV_RELATIVE)

            if (comment.id != commentId) {
                mBinding.divider.isInvisible = true
                displayLikeInfo()
            } else {
                mBinding.divider.isVisible = true
                mBinding.progressBar.isVisible = false
            }
        }

        private fun displayLikeInfo() {
            if (comment.isUploading() || comment.likeInProgress) {
                mBinding.imgLike.isInvisible = true
                mBinding.progressBar.isVisible = true
            } else {
                mBinding.imgLike.isVisible = true
                mBinding.progressBar.isVisible = false
                mBinding.imgLike.setImageResource(if (comment.isLiked) R.drawable.ic_favorite_red_24dp else R.drawable.ic_favorite_border_black_24dp)
            }
            if (comment.likeCount > 0) {
                mBinding.tvLikesCount.isVisible = true
                mBinding.tvLikesCount.text = "${comment.likeCount} ${if (comment.likeCount > 1) "likes" else "like"}"
            } else {
                mBinding.tvLikesCount.isVisible = false
            }
        }
    }

    private class ViewLoadingHolder(itemView: View) : BaseBindingViewHolder<ViewLoadingBinding, Void>(itemView) {

        override fun bindData(item: Void) {}

        override fun bindData(item: Void, payloads: List<Any>) {}
    }

    private class CommentDiffCallback : DiffUtil.ItemCallback<CommentUI>() {

        override fun areItemsTheSame(oldItem: CommentUI, newItem: CommentUI) = oldItem.id == newItem.id

        override fun areContentsTheSame(oldItem: CommentUI, newItem: CommentUI) = oldItem == newItem

        override fun getChangePayload(oldItem: CommentUI, newItem: CommentUI): Any? {
            val bundle = Bundle().apply {
                if (oldItem.likeInProgress != newItem.likeInProgress || oldItem.isLiked != newItem.isLiked)
                    putBoolean("like_status", true)
                if (oldItem.latestCommentId != newItem.latestCommentId)
                    putBoolean("latest_comment", true)
            }
            return if (!bundle.isEmpty) bundle else super.getChangePayload(oldItem, newItem)
        }
    }

    interface OnItemClickListener {

        fun onPhotoClicked(photo: Photo)

        fun onLikeClicked(commentId: String)
    }

    companion object {
        private const val VIEW_TYPE_COMMENT = 1
        private const val VIEW_TYPE_LOADING = 2
    }
}
